New JS interop docs (#4578)
dart-lang/site-www: src/interop/js-interop/js-types.md

---

> title: JS Types

Reviewer: Done!

description: Usage information about the core types in JS interop.
---

> ## Motivation

Reviewer: Done!

Dart values and JS values belong to separate domains. When compiling to Wasm,

> they belong in separate *runtimes* as well. As such, users should treat JS

Reviewer: Hmm, good points. I think we want the user to treat them as foreign types. Dart may or may not treat them as foreign types depending on the compiler, so we want users to play it safe and always treat them as foreign types. I updated some of the wording in other spots too so that it's a bit more clear.

> they belong in separate *runtimes* as well. As such, users should treat JS

Author: Interesting! Ok that makes sense, don't introduce the idea that they could do otherwise. Resolving this one

values as foreign types. In order to provide types for JS values, we use the set
of JS types available in `dart:js_interop`. These are extension types prefixed

> with `JS` and are called "JS types".

Reviewer: I largely agree with you on the first comment, although I like "To provide *Dart* types for JS values" better.

> similar to "you" vs "the user" in my previous comment, we usually default saying "Dart" or the specific actor (compiler, the package, etc.) instead of "we", or just attributing the action to the user if it's something "we" decided, but they actually do

Done (here and elsewhere)!.

I also like run on sentences (although I don't think qualifies :)), so done!

> with `JS` and are called "JS types".

Author: LGTM!

Importantly, these types have compiler-specific representation types. This means
that their runtime type will differ based on whether code is compiled to Wasm or
JS. In order to interact with and examine these JS values, we use `external`

> interop members.

Author: Probably want a link on both "representation types" and `external` for definitions, but like other links, that'll have to wait for more docs to form

Reviewer: Done!

> ## Type Hierarchy

Reviewer: Done!

JS types form a natural type hierarchy:

> 

Reviewer: I like it! If we're not going to link stuff directly, it might be better to make this a diagram, but I agree the search benefits are useful. I might come back to this - do we have goto program to create such a diagram?

> 

Author: I don't think we have an official go-to, but some recommendations from the team are:

- make it in Slides and do a screenshot
- Google Draw
- (highly recommended) Figma with the FigJam feature (less complicated than just Figma)
  - in this case you also need to provide the source for that diagram in the [diagrams folder](https://github.com/dart-lang/site-www/tree/main/diagrams)) using Omnigraffle (instructions are in the readme there). 
  - disclaimer: I've never done this myself

> 

Reviewer 2: As a note, any format is probably to put in the diagrams folder as long as it is loadable again and there is some breadcrumb to track what app to open it in (extension or commit message is fine).

I've also used https://app.diagrams.net/ for an experience like Google Draw but a bit more feature rich.

FigJam sounds interesting! I'll have to check it out :o

- Top type: `JSAny`, which is any non-nullish JS value
  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
  - `JSSymbol`
  - `JSBigInt`
  - `JSObject`, which is any JS object
    - `JSFunction`
      - `JSExportedDartFunction`, which represents a Dart callback that was
      converted to a JS function
    - `JSArray`
    - `JSPromise`
    - `JSDataView`
    - `JSTypedArray`
      - JS typed arrays e.g. `JSUint8Array`
    - `JSBoxedDartObject`, which allows users to box and pass Dart values
      opaquely within the same Dart runtime

Note that these JS types are *also* interop types. As such, users can and should
use them as representation types of other interface types. Users will likely

> want to use `JSObject` for this purpose.

Reviewer: > I'm not sure what this is trying to say. So far, it seems like the page defines JS types as a mechanism made up in dart:js_interop to...interop with JS types. So what's an "interop type" if not just another way to describe a JS type? Is it like, you use these same types/their same exact names in your interop code?

I was hoping to define "interop type" in the syntax section as it's a better fit, but I can see that this reads as super confusing without the additional context. My goal is to communicate some distinction between the types declared in `dart:js_interop` (which we refer to as JS types) and all interop types that you can use (which are transitively extension types on JS types/`@staticInterop` types/maybe `dart:html` types as well as `@staticInterop` types).

As an example:

```dart
import 'dart:js_interop';
import 'dart:html' show Window;

extension type InteropExtensionType(JSObject o) {}

extension type InteropExtensionType2(InteropExtensionType o) {}

extension type InteropExtensionType3(StaticInteropType o) {}

extension type InteropExtensionType4(Window o) {}

@JS()
@staticInterop
class StaticInteropType {}
```

JS types: `JSString`, `JSNumber`, all the other types in `dart:js_interop`
Interop types: `JSString`, `JSNumber`, all the other types in `dart:js_interop`, `InteropExtensionType`, `InteropExtensionType2`, `InteropExtensionType3`, `InteropExtensionType4`, `StaticInteropType`, `Window`

I *think* what I should do instead is remove these comments and have an "interop type" section in the syntax which refers to "JS types". In that same section, I'll likely want to mention why using `JSObject` makes sense.

> want to use `JSObject` for this purpose.

Author: Thanks for explaining! That makes a lot of sense and yeah sorry since this is all coming out non-linearly I should've expected that everything won't be spelled out here. 

Removing the text works fine. But also, with this context, I do actually like your original text now, since it would point to an explanation. But this part: 
> "...use them as representation types of other interface types. Users will likely
want to use `JSObject` for this purpose." 

...Would have to be repeated and elaborated on over there. Up to you; you can decide which is better.

Wherever you define interop types, though, I hope you use something like your example in the comment (plus the list separating which is which afterwards); that really helped me make sense of it!

> want to use `JSObject` for this purpose.

Reviewer: > ...Would have to be repeated and elaborated on over there. Up to you; you can decide which is better.

Good point. I think I'll wait and see when I write the section on syntax - it may end up that here is the better fit for that context.

Noted on using an example/better differentiating JS types and interop types!

// TODO(srujzs): Add a link to the interop type section in the syntax.
// TODO(srujzs): Should we link the type to the definition? It might be a bit

> // annoying to do everywhere a JS type is mentioned.

Reviewer: Thanks for the comment syntax - I could never tell if markdown had a canonical way to comment. I like the idea of linking to the `dart:js_interop` homepage! It avoids having to spell out every single typed array type (which I think would clutter the page).

I think we actually want to link to the latest documentation here, as these are extension types now so the link we might want is https://api.dart.dev/dev/dart-js_interop/dart-js_interop-library.html#extension-types. Alternatively, we can use beta. Of course we can keep it as stable, but the link won't work until 3.3 is released. Thoughts?

> // annoying to do everywhere a JS type is mentioned.

Author: Oh right, I'm used to waiting until the stable release to merge documentation. But yeah if your goal is just to publish all this as soon as it's ready, so possibly before the release, then yes definitely use dev!

Prefer to only convert when you need to. Furthermore, conversions may or may not
produce a new value. This doesn’t matter for immutable values like numbers, but
does matter for types like `List`s. A conversion to a `JSArray` may produce a
new value by copying or may not, so do not rely on modifications to the
`JSArray` affecting the `List`. Typed array conversions have a similar
limitation. Look at the specific conversion function for more details.

## Requirements on `external` declarations and `Function.toJS`

In order to ensure type safety and consistency, we place requirements on what
types can flow into and out of JS. Passing arbitrary Dart values into JS is not
allowed. Instead, we require users to use a compatible interop type like a JS
type or a primitive, which would then be implicitly converted by the compiler.
For example, these would be allowed:

```dart
@JS()
external void usePrimitives(String a, int b, double c, num d, bool e);
```
```dart
@JS()
external JSArray useJsTypes(JSObject _, JSString __);
```
```dart
extension type InteropType(JSObject _) {}

@JS()
external InteropType get interopType;

> ```

Reviewer: That's `good` to know, thanks!

whereas these would return an error:

```dart
@JS()
external Function get function;
```

```dart
@JS()
external set list(List _);

> ```

Reviewer: Done!

These same requirements exist when we use `Function.toJS` to make a Dart
function callable in JS. The values that flow into and out of this callback must
be a compatible interop type or a primitive.

If you use a Dart primitive e.g. `String`, an implicit conversion happens in the
compiler to convert that value from a JS value to a Dart value. If performance
is critical and you don’t need to examine the contents of the string, then maybe
using `JSString` instead makes sense to avoid the conversion cost.

> ## Compatibility/type checks and casts

Reviewer: Done!

JS has both a `null` and an `undefined` value. This is in contrast with Dart,
which only has `null`. In order to make JS values more ergonomic to use, if an
interop member were to return either JS `null` or `undefined`, we map these
values to Dart `null`. Therefore a member like:

```dart
@JS()
external JSObject? get value;
```

can be interpreted as returning a JS object, JS `null`, or `undefined`.

**Important**: There is a subtle inconsistency with regards to `undefined`
between compiling to JS and Wasm. While compiling to JS treats `undefined`
values like Dart `null`, it doesn’t actually *change* the value itself. If you
an interop member returns `undefined` and you pass it back into JS, JS will see
`undefined` and not `null` when compiling to JS. When compiling to Wasm, this is
not the case, because we convert the value to Dart `null`, thereby losing
information on whether the original value was JS `null` or `undefined`. Avoid
writing code where this distinction matters by explicitly passing Dart `null`
instead to an interop member. Currently, there’s no platform-consistent way to
provide `undefined` to interop members or distinguish between JS `null` and

> `undefined` values, but this will likely change in the future.

Reviewer: I think the first two should point to the extension types documentation. `external` is interesting, it looks like https://dart.dev/language/keywords doesn't have a dart.dev link for it but rather points directly to the spec. I'll link to the keywords page here instead so it doesn't point to the spec if that ever changes.

> `undefined` values, but this will likely change in the future.

Author: Huh, look at that. We'll definitely be talking about external a lot between js interop and extension types, so I created an [issue](https://github.com/dart-lang/site-www/issues/5478) to add a dart.dev section for it. Thanks for bringing it to my attention!

    - `JSArray`
    - `JSPromise`
    - `JSDataView`
    - `JSTypedArray`
      - JS typed arrays e.g. `JSUint8Array`
    - `JSBoxedDartObject`, which allows users to box and pass Dart values
      opaquely within the same Dart runtime

You can find the definition of each type in the [`dart:js_interop` API docs].

{% comment %}
TODO (srujzs): Refer to this in the "interop type" section in syntax.
TODO (srujzs): Should we add a tree diagram instead for JS types?
{% endcomment %}

## Conversions

Since there are two separate domains, you will likely want to *convert* values
between one domain to the other. For example, you may want to convert a Dart
`String` into a JS string, which is represented by the JS type `JSString`. To do
this, Dart supplies a number of `extension` members on various Dart types that
will do the conversion for you, usually labeled as or prefixed with `toJS`.
Similarly, you can convert JS values to Dart types. Dart supplies a number of
members on various JS types to enable this conversion, usually labeled as or
prefixed with `toDart`. They are often marked `external`, and their
implementation is usually platform-dependent. Note that not all JS types have a

> conversion, and not all Dart types have a conversion.

Reviewer: Thanks! My modifications after incorporating yours:

> to get the behavior and output you're expecting.

I think this part will likely confuse users more as it doesn't make clear what behavior and output they expect.

> This rewrite just adds a clause again just adding some context of why this is important.

I like the point of adding why users may want to do this, but ironically, `JSString` is a bad candidate since we allow you to write `String` on interop APIs. :) So, I opted to make this `List` <-> `JSArray` so that the motivation is more realistic.

> (or, to abstract the complexity of conversions for you)

While this is true for some conversions, users can't write the conversions themselves in many cases as they are implementation details that users don't have access to.

> If it's the latter, does the reader need to know about their implementation? If there's something the user needs to do when using these members that is dependent on whether the member's implementation marks it as external or not, can you explain why they need to know that here?

It's a good question. `external` is an overloaded keyword that serves many purposes, and I'm trying to reduce the confusion between the `external` members users can write and the ones that are written in the SDK. Some of the latter are *patched* members and not interop members i.e. they are implemented based on the compiler and don't have anything to do with JS interop. This may be a bit too much info for users so I shortened it to just mention that users shouldn't mistake this with interop members.

In general, the conversion table looks like the following:

| JS type                             | Dart type                                |
| ----------------------------------- | ---------------------------------------- |
| `JSNumber`, `JSBoolean`, `JSString` | `num`, `int`, `double`, `bool`, `String` |
| `JSExportedFunction`                | `Function`                               |
| `JSArray<T extends JSAny?>`         | `List<T extends JSAny?>`                 |
| `JSPromise<T extends JSAny?>`       | `Future<T extends JSAny?>`               |
| Typed arrays e.g. `JSUint8Array`    | `dart:typed_data`                        |

> | `JSBoxedDartObject`                 | Opaque Dart value                        |

Reviewer: Done!

**Important**: Conversions may have different costs depending on the compiler.
Prefer to only convert when you need to. Furthermore, conversions may or may not
produce a new value. This doesn’t matter for immutable values like numbers, but
does matter for types like `List`s. A conversion to a `JSArray` may produce a
new value by copying or may not, so do not rely on modifications to the
`JSArray` to affect the `List`. Typed array conversions have a similar

> limitation. Look at the specific conversion function for more details.

Reviewer: Nice! Hopefully that makes it more obvious to users that it comes with a cost.

> I changed the "prefer" statement to try to make it a little clearer that the "when you need to" would be "when you can't make your program work otherwise", but idk if that's better or worse. Feel free to change back/change to something else!

Hmm, I too don't know if it's better or worse. :D I suppose when in doubt, choose brevity so I'll keep that part as is.

> Also, when you say "this doesn't matter for immutable values like numbers", does that mean it won't ever affect them (i.e. an immutable value will never be changed by a conversion)?

More that it doesn't *matter* whether you get a new value or not, because you can't change the contents of immutable values. You can't rely on modifications of an immutable value to affect the converted value even if you wanted to, because they can't be modified.

> ## Requirements on `external` declarations and `Function.toJS`

Author: Is this related to the previous section, like kind of a subtopic? If son
```suggestion
### Requirements on `external` declarations and `Function.toJS`
```

Reviewer: I think it's a separate topic. Conversions are related, but not all of the information in this sections has to do with conversions.

In order to ensure type safety and consistency, the compiler places requirements
on what types can flow into and out of JS. Passing arbitrary Dart values into JS
is not allowed. Instead, the compiler requires users to use a compatible interop
type like a JS type or a primitive, which would then be implicitly converted by
the compiler. For example, these would be allowed:

{:.good}
```dart
@JS()
external void primitives(String a, int b, double c, num d, bool e);
```

{:.good}
```dart
@JS()
external JSArray jsTypes(JSObject _, JSString __);
```

{:.good}
```dart
extension type InteropType(JSObject _) {}

@JS()
external InteropType get interopType;
```

> whereas these would return an error:

Author: ```suggestion
Whereas these would return an error:
```

Reviewer: Done

{:.bad}
```dart
@JS()
external Function get function;
```

{:.bad}
```dart
@JS()
external set list(List _);
```

These same requirements exist when you use [`Function.toJS`] to make a Dart
function callable in JS. The values that flow into and out of this callback must
be a compatible interop type or a primitive.

> If you use a Dart primitive e.g. `String`, an implicit conversion happens in the

Author: ```suggestion
If you use a Dart primitive like `String`, an implicit conversion happens in the
```

We're supposed to avoid using e.g./i.e. in text, I think for translation or accessibility purposes

Reviewer: Good to know! I have no complaints because `like` is the exact same number of characters as `e.g.` and `i.e.` so I don't have to deal with line wrapping here.

compiler to convert that value from a JS value to a Dart value. If performance
is critical and you don’t need to examine the contents of the string, then using

> `JSString` instead to avoid the conversion cost may make sense.

Author: ````suggestion
`JSString` instead to avoid the conversion cost might make sense:

```dart
```
````

Would it make sense to show a small example of what you mean by using `JSString` instead? Or is that basically what you've already shown in the second `good` example?

Reviewer: Yup! I mentioned the second example to be more explicit.

## Compatibility, type checks, and casts

> As mentioned above, the representation type of JS types may differ based on the
> compiler. This affects runtime type-checking and casts. Therefore, almost always
> avoid `is` checks where the value is an interop type or where the target type is
> an interop type. Avoid casts between Dart types and interop types. In order to
> type-check a JS value, use an interop member like [`typeofEquals`] or
> [`instanceOfString`] that examines the JS value itself.
> 
> Bad:
> 
> {:.bad}
> ```dart
> void f(JSAny a) {
>   if (a is String) { … }
> }
> ```
> 
> {:.bad}
> ```dart
> void f(JSObject o) {
>   if (o is JSObject) { … }
> }
> ```
> 
> {:.bad}
> ```dart
> void f(JSString s) {
>   s as String;
> }
> ```
> 
> Okay:

Author: Removed "As mentioned..." since people might end up here following a link directly to this section (and the above mention is pretty light anyway)

I didn't rewrite anything else, just moved the sentences describing each situation to right about the code block showing that situation (so you don't have to say good/bad)

````suggestion
The representation type of JS types may differ based on the compiler.
This affects runtime type-checking and casts. Therefore, almost always
avoid `is` checks where the value is an interop type or where the target type is
an interop type:

{:.bad}
```dart
void f(JSAny a) {
  if (a is String) { … }
}
```

{:.bad}
```dart
void f(JSObject o) {
  if (o is JSObject) { … }
}
```

Also, avoid casts between Dart types and interop types:

{:.bad}
```dart
void f(JSString s) {
  s as String;
}
```

To type-check a JS value, use an interop member like [`typeofEquals`] or
[`instanceOfString`] that examines the JS value itself:
````

Reviewer: I actually just landed a better way for users to do this, so I'll add a TODO to link that instead once it's in a dev release. :D Users won't have access to it until 3.4, so I'll still mention the other two and add examples.

{:.good}
```dart
void f(JSAny a) {
  // Here `a` is verified to be a JS function, so the cast is okay.
  if (a.typeofEquals('function')) {
    a as JSFunction;
  }
}
```

> Dart may add lints to make runtime checks with JS interop types easier to avoid.

Author: If there's an issue where people can go vote on this, you should link it here!

Reviewer: Done!

## `null` vs `undefined`

JS has both a `null` and an `undefined` value. This is in contrast with Dart,
which only has `null`. In order to make JS values more ergonomic to use, if an
interop member were to return either JS `null` or `undefined`, the compiler maps

> these values to Dart `null`. Therefore a member like:
> 
> ```dart
> @JS()
> external JSObject? get value;
> ```
> 
> can be interpreted as returning a JS object, JS `null`, or `undefined`.

Author: The example in the middle of the sentence is kind of weird, so no big deal but if it makes sense try:

````suggestion
these values to Dart `null`. Therefore a member like `value` in the following example
can be interpreted as returning a JS object, JS `null`, or `undefined`:

```dart
@JS()
external JSObject? get value;
```
````

Is `value` even the member you're referring to? 😅  Whatever it is, hopefully you get the idea!

Reviewer: It is, so this is clearer, thanks!

> **Important**: There is a subtle inconsistency with regards to `undefined`
> between compiling to JS and Wasm. While compiling to JS *treats* `undefined`
> values as if they were Dart `null`, it doesn’t actually *change* the value
> itself. If an interop member returns `undefined` and you pass that value back
> into JS, JS will see `undefined` and not `null` when compiling to JS.
> However, when compiling to Wasm, this is not the case and the value will be
> `null` in JS. This is because the compiler *converts* the value to Dart `null`
> when compiling to Wasm, thereby losing information on whether the original value
> was JS `null` or `undefined`. Avoid writing code where this distinction matters
> by explicitly passing Dart `null` instead to an interop member. Currently,
> there's no platform-consistent way to provide `undefined` to interop members or
> distinguish between JS `null` and `undefined` values, but this will likely
> change in the future.

Author: Didn't rewrite anything, just added the site alert and broke up the paragraphs:

```suggestion
{{site.alert.warn}}
There is a subtle inconsistency with regards to `undefined`
between compiling to JS and Wasm. While compiling to JS *treats* `undefined`
values as if they were Dart `null`, it doesn’t actually *change* the value
itself. If an interop member returns `undefined` and you pass that value back
into JS, JS will see `undefined`, *not* `null`, when compiling to JS.

However, when compiling to Wasm, this is not the case and the value will be
`null` in JS. This is because the compiler *converts* the value to Dart `null`
when compiling to Wasm, thereby losing information on whether the original value
was JS `null` or `undefined`. Avoid writing code where this distinction matters
by explicitly passing Dart `null` instead to an interop member. 

Currently, there's no platform-consistent way to provide `undefined` to
interop members or distinguish between JS `null` and `undefined` values,
but this will likely change in the future.
{{site.alert.end}}
```

The part where it says "The compiler converts the value to Dart `null`...", would that be an _implicit conversion_?
Would it add anything to add "implicit" there? Just wondering since you talk about implicit conversions earlier it would be nice to make that call back, and since you mention explicit later here. 

If there's an issue where the "likely change in the future" is being tracked, link that here too!

Reviewer: > would that be an implicit conversion? Would it add anything to add "implicit" there?

Indeed! This is a good example of an implicit conversion versus the explicit conversions users can do themselves.

Done wrt adding the issue.

to a JS interop API.

Dart supplies a number of conversion members on various Dart types and JS types
to convert the values between the domains for you.

Members that convert values from Dart to JS usually start with `toJS`:

```dart
String str = 'hello world';
JSString jsStr = str.toJS;
```

Members that convert values from JS to Dart usually start with `toDart`:

```dart
JSNumber jsNum = ...;
int integer = jsNum.toDartInt;
```

Not all JS types have a conversion, and not all Dart types have a conversion.
Generally, the conversion table looks like the following:

<div class="table-wrapper" markdown="1">
| JS type                             | Dart type                                |
| ----------------------------------- | ---------------------------------------- |
| `JSNumber`, `JSBoolean`, `JSString` | `num`, `int`, `double`, `bool`, `String` |
| `JSExportedFunction`                | `Function`                               |
| `JSArray<T extends JSAny?>`         | `List<T extends JSAny?>`                 |
| `JSPromise<T extends JSAny?>`       | `Future<T extends JSAny?>`               |
| Typed arrays like `JSUint8Array`    | `dart:typed_data`                        |

> | `JSBoxedDartObject`                 | Opaque Dart value                        |

Reviewer: I like specifying typed arrays, but disagree with `Object` as it misleads the purpose the conversion. I'm worried that might encourage people to convert everything using `toJSBox`.

{:.table .table-striped}
</div>

{{site.alert.warn}}

> Conversions might have different costs depending on the compiler, so prefer to

Reviewer: Done.

only convert values if you need to. Conversions also may or may not produce a
new value. This doesn’t matter for immutable values like numbers, but does
matter for types like `List`. A conversion to a `JSArray` may produce a new
value by copying or may not, so do not rely on modifications to the `JSArray` to

> affect the `List`. Typed array conversions have a similar limitation. Look up

Reviewer: Cleared up what the users could expect with a list conversion.

---
title: JS types
description: Usage information about the core types in JS interop.
---

Dart values and JS values belong to separate language domains. When compiling to
Wasm, they execute in separate *runtimes* as well. As such, you should treat JS
values as foreign types. To provide Dart types for JS values,

> [`dart:js_interop`] exposes a set of [extension types] prefixed with `JS` called

Reviewer 3: nit: I think the fact that they are extension-types are an implementation detail that we can delay to explain later. Connecting this with the comment below, I think if we leave out the "representation type" detail from the next paragraph, but explain it more with the lens of them being inconsistent and unreified, then we can probably also drop here the fact that they are currently implemented as extension-types.

Reviewer: Agreed that we can remove the notion of extension types and representation types and just refer to inconsistent runtime types.

"JS types".

> Importantly, these types have compiler-specific [representation types]. This

Reviewer 3: Consider shuffling how we present the info in this paragraph, to be more upfront about what this means for the reader/developer, and if we find that it's necessary we can explain the rationale.

I think that the "representation types" is the reason behind things, but what matters to developers is that these JS types are not really reified at runtime and not even consistent across implementations.

A possible suggestion:
> JS types are used to help developers and tools track, at compile-time, when the program is operating over Dart or JS values. However, at runtime, JS types are not fully reified and their implementation is different when you compile to Wasm or JS. This imposes important [restrictions](#restrictions) of how to use these values: they cannot be used in is-tests and casts like regular Dart types. You can, however, work with them through conversions or through interop calls.

Reviewer: Reworded to take into account your feedback, highlighting the motivation, the difference in runtime semantics, and the possibility of using conversions.

use [`external`] interop members.

## Type hierarchy

JS types form a natural type hierarchy:

- Top type: `JSAny`, which is any non-nullish JS value
  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
  - `JSSymbol`
  - `JSBigInt`
  - `JSObject`, which is any JS object
    - `JSFunction`
      - `JSExportedDartFunction`, which represents a Dart callback that was
      converted to a JS function
    - `JSArray`
    - `JSPromise`
    - `JSDataView`
    - `JSTypedArray`
      - JS typed arrays like `JSUint8Array`
    - `JSBoxedDartObject`, which allows users to box and pass Dart values
      opaquely within the same Dart runtime

You can find the definition of each type in the [`dart:js_interop` API docs].

{% comment %}
TODO (srujzs): Should we add a tree diagram instead for JS types?
{% endcomment %}

## Conversions

To use a value from one domain to another, you will likely want to *convert* the
value to the corresponding type of the other domain. For example, you might want
to convert a Dart `List<JSString>` into a JS array of strings, which is
represented by the JS type `JSArray<JSString>`, so that you can pass the array
to a JS interop API.

> Dart supplies a number of `extension` members on various Dart types and JS types
> to convert the values between the domains for you.

Reviewer 3: Small consideration: maybe we can talk about the fact that they are extension members later. 

For example:
```suggestion
Dart provides a wide variety of members to help you convert values between the two domains.
```

Then when describing the `.toJS` name we can also share that they are added on as extension members. I debate whether we should also say it for `.toDart` (since the member and the type are exposed together, and I could see us one day moving the `.toDart` into the extension type).

For example:
> Helper members to convert from Dart to JS are exposed as extension members on Dart types with names starting with `.toJS`: ...sample-code...
> Conversely, to convert from JS to Dart, you can use members with names starting with `.toDart` on JS types.

Reviewer: I think it's a good idea to omit the fact that they are extension members altogether. Mentioning it for the `toJS` case but not the `toDart` case makes it a little confusing as today they're extension members as well.

(I can totally see the motivation for mentioning it though. Trying to use conversion functions with just an import to `package:web` resulting in an error may seem initially confusing.)

Members that convert values from Dart to JS are usually labeled-as or

> prefixed-with `toJS`:

Reviewer: Done!

```dart
String str = 'hello world';
JSString jsStr = str.toJS;
```

Members that convert values from JS to Dart are usually labeled-as or
prefixed-with `toDart`:

```dart
JSNumber jsNum = ...;
int integer = jsNum.toDartInt;
```

> These conversion functions are often marked `external`, but aren't interop

Author: There's an [open PR](https://github.com/dart-lang/site-www/pull/5481) right now for defining `external` in the docs (I'll merge that soon). I can add these details to that, and then can we just link to that section here? That could help avoid going into too much detail here.

I support dropping it just because the example code right above doesn't show it with `external`, so that was just confusing for me to read. If it's the more common scenario, maybe put `external` in the example code?

> These conversion functions are often marked `external`, but aren't interop

Reviewer: My general concern was that users would look at the API surface in js_interop.dart, see that there are interop members on JS types that are marked `external`, but then also see that other members are marked as `external` that don't follow the same semantics. Maybe the better way to solve this is to add/require the `@patch` annotation on the user-facing declaration so it's obvious that it's different from other `external` members.

I'm okay with removing this as it doesn't really dive into that difference anyways, so only half-explaining it is possibly worse than not mentioning it.

> doesn't show it with external, so that was just confusing for me to read. If it's the more common scenario, maybe put external in the example code?

The conversion functions being used are `external` e.g. `String.toJS`, but if we're highlighting their usage, the example wouldn't show the `external` as that's tied to the declaration. Maybe linking to the declaration would reduce the confusion, but this sentence is being removed, so nevermind. :)

```

{:.good}
```dart
extension type InteropType(JSObject _) {}

@JS()
external InteropType get interopType;
```

Whereas these would return an error:

{:.bad}
```dart
@JS()
external Function get function;
```

{:.bad}
```dart
@JS()
external set list(List _);
```

These same requirements exist when you use [`Function.toJS`] to make a Dart
function callable in JS. The values that flow into and out of this callback must
be a compatible interop type or a primitive.

If you use a Dart primitive like `String`, an implicit conversion happens in the
compiler to convert that value from a JS value to a Dart value. If performance
is critical and you don’t need to examine the contents of the string, then using
`JSString` instead to avoid the conversion cost might make sense like in the
second example.

## Compatibility, type checks, and casts

The representation type of JS types may differ based on the compiler. This
affects runtime type-checking and casts. Therefore, almost always avoid `is`
checks where the value is an interop type or where the target type is an interop
type:

{:.bad}
```dart
void f(JSAny a) {
  if (a is String) { … }
}
```

> {:.bad}

Reviewer 3: for some reason this renrered normal, maybe it needs an empty line in between?

Author: Probably, I added a suggestion comment to fix it

Reviewer: Done!

> ```
> {:.bad}

Author: ````suggestion
```

{:.bad}
````

Reviewer: Done!

```dart

> void f(JSObject o) {

Reviewer 3: did you mean to use JSAny here (otherwise it's the same type as the `is` check)

Reviewer: Yes, good catch!

JS has both a `null` and an `undefined` value. This is in contrast with Dart,
which only has `null`. In order to make JS values more ergonomic to use, if an
interop member were to return either JS `null` or `undefined`, the compiler maps
these values to Dart `null`. Therefore a member like `value` in the following
example can be interpreted as returning a JS object, JS `null`, or `undefined`:

```dart
@JS()
external JSObject? get value;
```

> If the return type was not declared as nullable, then the compiler will throw an

Reviewer 3: nit: s/compiler/program? (technically the compiler will complete normally, but the output program will throw :smile: )

Reviewer: Yes, thanks :)
